module/zfs/vdev_raidz_math.c

```diff
@@ -0,0 +1,713 @@
+			return (rm->rm_ops->rec[RAIDZ_REC_Q]);
+		} else if (parity_valid[CODE_R]) {
+			return (rm->rm_ops->rec[RAIDZ_REC_R]);
+		}
+	} else if (nbaddata == 2) {
+		if (parity_valid[CODE_P] && parity_valid[CODE_Q]) {
+			return (rm->rm_ops->rec[RAIDZ_REC_PQ]);
+		} else if (parity_valid[CODE_P] && parity_valid[CODE_R]) {
+			return (rm->rm_ops->rec[RAIDZ_REC_PR]);
+		} else if (parity_valid[CODE_Q] && parity_valid[CODE_R]) {
+			return (rm->rm_ops->rec[RAIDZ_REC_QR]);
+		}
+	} else if (nbaddata == 3 &&
+		parity_valid[CODE_P] && parity_valid[CODE_Q] &&
+		parity_valid[CODE_R]) {
+		return (rm->rm_ops->rec[RAIDZ_REC_PQR]);
+	}
+	return ((raidz_rec_f) NULL);
+}
+
+/*
+ * Select data reconstruction method for raidz_map
+ * @parity_valid - Parity validity flag
+ * @dt           - Failed data index array
+ * @nbaddata     - Number of failed data columns
+ */
+int
+vdev_raidz_math_reconstruct(raidz_map_t *rm, const int *parity_valid,
+	const int *dt, const int nbaddata)
+{
+	raidz_rec_f rec_data = NULL;
+
+	switch (raidz_parity(rm)) {
+		case 1:
+			rec_data = _reconstruct_fun_raidz1(rm, parity_valid,
+			    nbaddata);
+			break;
+		case 2:
+			rec_data = _reconstruct_fun_raidz2(rm, parity_valid,
+			    nbaddata);
+			break;
+		case 3:
+			rec_data = _reconstruct_fun_raidz3(rm, parity_valid,
+			    nbaddata);
+			break;
+		default:
+			cmn_err(CE_PANIC, "invalid RAID-Z configuration %d",
+			    raidz_parity(rm));
+			break;
+	}
+
+	ASSERT(rec_data != NULL);
+
+	return (rec_data(rm, dt));
+}
+
+static void
+init_raidz_kstat(raidz_math_ops_kstat_t *rs, const char *name)
+{
+	int i;
+	const size_t impl_name_len = strnlen(name, KSTAT_STRLEN);
+	const size_t op_name_max = (KSTAT_STRLEN-2) > impl_name_len ?
+		KSTAT_STRLEN - impl_name_len - 2 : 0;
+
+	for (i = 0; i < RAIDZ_GEN_NUM; i++) {
+		strncpy(rs->gen_kstat[i].name, name, impl_name_len);
+		strncpy(rs->gen_kstat[i].name + impl_name_len, "_", 1);
+		strncpy(rs->gen_kstat[i].name + impl_name_len + 1,
+			raidz_gen_name[i], op_name_max);
+
+		rs->gen_kstat[i].data_type = KSTAT_DATA_UINT64;
+		rs->gen_kstat[i].value.ui64  = 0;
+	}
+
+	for (i = 0; i < RAIDZ_REC_NUM; i++) {
+		strncpy(rs->rec_kstat[i].name, name, impl_name_len);
+		strncpy(rs->rec_kstat[i].name + impl_name_len, "_", 1);
+		strncpy(rs->rec_kstat[i].name + impl_name_len + 1,
+			raidz_rec_name[i], op_name_max);
+
+		rs->rec_kstat[i].data_type = KSTAT_DATA_UINT64;
+		rs->rec_kstat[i].value.ui64  = 0;
+	}
+}
+
+
+/* Number of pages for benchmark: 2^BENCH_SIZE_ORD */
+#define	BENCH_D_COLS	(8)
+#define	BENCH_COLS	(3 + BENCH_D_COLS)
+#define	BENCH_COL_SIZE	(512)
+#define	BENCH_ZIO_SIZE	(BENCH_COL_SIZE * BENCH_D_COLS)
+#define	BENCH_NS	(1ULL*1000ULL*1000ULL) /* 50ms */
+
+#define	BENCH_RES_FMT	"RAIDZ_math: [%-7s] %-8s %5ld.%.3ld MB/s\n"
+#define	BENCH_SKIP_FMT	"RAIDZ_math: [%-7s] is not supported.\n"
+
+#if defined(_KERNEL)
+#define	bench_print(x...)	cmn_err(CE_CONT, x)
+#else
+#define	bench_print(x...)	do { } while (0)
+#endif
```

Comment:
Personally I don't think here's much benefit in printing this to the console, having access to this through the kstat seems sufficient to me so I'd suggest dropping `bench_print()` entirely.